util/sshutil/scpurl.go

```diff
@@ -2,14 +2,18 @@ package sshutil
 
 import (
 	"errors"
-	"fmt"
 	"net/url"
 	"regexp"
 )
 
-var gitSSHRegex = regexp.MustCompile("^([a-zA-Z0-9-_]+)@([a-zA-Z0-9-.]+):(.*?)(?:#(.*))?$")
+var gitSSHRegex = regexp.MustCompile("^(?:([a-zA-Z0-9-_]+)@)?([a-zA-Z0-9-.]+):(.*?)(?:#(.*))?$")
```

Comment:
I wonder if this can cause false positives.

Possibilities:
- What if we require `.` in hostname
- Would requiring `.git` in repo be too awful?

Reply:
Mm, it definitely can cause false positives.

The problem with introducing those additional constraints, is this would be a potentially breaking change. We don't currently require `.git`, or require a hostname with a `.` in it. We could make this change using a capability, but the `.git` change definitely feels like it could break things.

However, because of the `NoImplicitSCPUsername` option which gets called for `ParseGitRef`, this change won't affect anything that's dockerfile-facing. It's only in the LLB layer, when we've passed a URL to `llb.Git`, and so have already decided it's a git url (so a custom frontend that has a way of indicating that this is a git url, e.g. `dag.Git("<url>")` in dagger can handle these URL types).

Reply:
In the dockerfile case we would already decide if we call `llb.HTTP` or `llb.Git`. If we choose git then it can't be anything else.

Reply:
Right - I think the regexp can *potentially* cause false positives, but I don't see the issue: buildkit behavior continues working as before (except some previously invalid calls to `llb.Git` are now valid), and the regexp now expands to match "the spec" (and more importantly the `git` cli).